src/org/jointheleague/level4/minesweeper/MineSweeper.java

```diff
@@ -0,0 +1,132 @@
+     */
+    private void resetGame(Object unused) {
+        // TODO fill in
+    }
+    
+    private void createAndShowFrame() {
+        final JMenuItem resetMenuItem = new JMenuItem("Reset");
+        resetMenuItem.addActionListener(null); // TODO replace null with method reference that resets the game
+        
+        final JMenu gameMenu = new JMenu("Game");
+        gameMenu.add(resetMenuItem);
+        
+        final JMenuBar menuBar = new JMenuBar();
+        menuBar.add(gameMenu);
+        
+        final JPanel controlPanel = new JPanel();
+        final JButton resetButton = new JButton("Reset");
+        resetButton.addActionListener(null); // TODO replace null with method reference that resets the game
+        controlPanel.add(resetButton);
+        
+        final JPanel gameBoardPanel = new JPanel();
+        gameBoardPanel.setSize(WIDTH * CELL_SIZE, HEIGHT * CELL_SIZE);
+        gameBoardPanel.setBackground(Color.WHITE);
+        gameBoardPanel.setLayout(new GridLayout(HEIGHT, WIDTH));
+        IntStream.range(0, HEIGHT).forEach(y ->
+            IntStream.range(0, WIDTH).forEach(x -> {
+                // This code loops through the X and Y indexes,
+                // creating a button for each cell.
+                final JButton b = new JButton();
+                buttons[y][x] = b;
+                // When the cell button is pressed, it should:
+                // 1. Initializes the mines if it is not yet initialized.
+                // 2. If the button is a mine:
+                //    - Change the button text to an "X"
+                //    - Display "You Lose" in a dialog box
+                //    Otherwise:
+                //    - Change the button text to the number of neighboring
+                //      cells are mines.
+                //    - Decrement `numCellsRemaining`
+                //    - If all cells are open, display "You Win" in a dialog box
+                //    - Extra credit: If the number of neighboring cells is 0,
+                //      automatically open all neighboring cells
+                b.addActionListener(null); // TODO replace null with lambda expression
+                gameBoardPanel.add(b);
+            })
+        );
+        
+        final JFrame frame = new JFrame();
+        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
+        frame.setSize(WIDTH * CELL_SIZE, (HEIGHT + 1) * CELL_SIZE + 27);
+        frame.setLayout(new BorderLayout());
+        frame.setResizable(false);
+        frame.setJMenuBar(menuBar);
+        frame.add(gameBoardPanel, BorderLayout.CENTER);
+        frame.add(controlPanel, BorderLayout.PAGE_END);
+        
+        frame.setVisible(true);
+    }
+    
+    public static void main(String[] args) {
+        System.setProperty("apple.laf.useScreenMenuBar", "true");
+        final MineSweeper mineSweeper = new MineSweeper();
+        SwingUtilities.invokeLater(mineSweeper::createAndShowFrame);
+    }
+}
```

Comment:
Add tests so that the students can test their code as they progress. This gets them used to TDD and gives them a few carrots along the way.